torch/testing/_internal/common_device_type.py

```diff
@@ -693,6 +694,14 @@ def get_desired_device_type_test_bases(except_for=None, only_for=None, include_l
     def split_if_not_empty(x: str):
         return x.split(",") if x else []
 
+    # run some cuda testcases on other devices if available
+    # Usage:
+    # export PYTORCH_TESTING_DEVICE_FOR_CUSTOM=privateuse1
+    env_custom_only_for = split_if_not_empty(os.getenv(PYTORCH_TESTING_DEVICE_FOR_CUSTOM_KEY, ''))
+    if env_custom_only_for:
+        desired_device_type_test_bases += filter(lambda x: x.device_type in env_custom_only_for, test_cases)
```

Comment:
What exactly is going on with the filter here?

Reply:
Like the example: instantiate_device_type_tests(TestAssertCloseMultiDevice, globals(), only_for="cuda").
run the testcases on the **privateuse1** device.
the debugging code is:
```
def get_desired_device_type_test_bases(except_for=None, only_for=None, include_lazy=False, allow_mps=False):
    test_bases = device_type_test_bases.copy()
    if allow_mps and TEST_MPS and MPSTestBase not in test_bases:
        test_bases.append(MPSTestBase)

    # >>>> only_for is 'cuda'

    desired_device_type_test_bases = filter_desired_device_types(test_bases, except_for, only_for)

    # >>>> now desired_device_type_test_bases is []

    if include_lazy:
        if IS_FBCODE:
            print("TorchScript backend not yet supported in FBCODE/OVRSOURCE builds", file=sys.stderr)
        else:
            desired_device_type_test_bases.append(LazyTestBase)
    def split_if_not_empty(x: str):
        return x.split(",") if x else []
    env_custom_only_for = split_if_not_empty(os.getenv(PYTORCH_TESTING_DEVICE_FOR_CUSTOM_KEY, ''))

    # >>>> if env_custom_only_for is 'privateuse1'

    if env_custom_only_for:
        desired_device_type_test_bases += filter(lambda x: x.device_type in env_custom_only_for, test_cases)

    # >>>> now desired_device_type_test_bases is [<class 'torch.testing._internal.common_device_type.Privateuse1TestBase'>]

    env_only_for = split_if_not_empty(os.getenv(PYTORCH_TESTING_DEVICE_ONLY_FOR_KEY, ''))
    env_except_for = split_if_not_empty(os.getenv(PYTORCH_TESTING_DEVICE_EXCEPT_FOR_KEY, ''))
    return filter_desired_device_types(desired_device_type_test_bases, env_except_for, env_only_for)

```